### PR TITLE
Test: Implemented ginkgo test against local kubeconfig

### PR DESCRIPTION
**Pull request by ellistarn**
Fixes # 

**Description**
Enables e2e tests against a local cluster. 
- test resources are all tagged with a special label for cleanup
- Made progress on CI infra setup (critical addons only, role name, etc), but gaps remain

```
 ~/w/g/s/g/aws/karpenter │ on testinfra !1  make e2etests                                2 ✘ │ at 18:00:08
ginkgo -r test -- --environment-name=karpenter-test-infrastructure
Running Suite: Integration Suite
================================
Random Seed: 1656550823
Will run 1 of 1 specs

• [SLOW TEST:85.879 seconds]
Sanity Checks
/Users/etarn/workspaces/go/src/github.com/aws/karpenter/test/suites/integration/suite_test.go:30
  should provision nodes
  /Users/etarn/workspaces/go/src/github.com/aws/karpenter/test/suites/integration/suite_test.go:31
------------------------------

Ran 1 of 1 Specs in 87.084 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 0 Skipped
PASS

Ginkgo ran 1 suite in 1m33.141041379s
Test Suite Passed
```

**How was this change tested?**

*

**Does this change impact docs?**
- [ ] Yes, PR includes docs updates
- [ ] Yes, issue opened: # 
- [ ] No

**Release Note**

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: test/pkg/environment/expectations.go [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+		&v1.PersistentVolumeClaim{},
+		&v1.PersistentVolume{},
+		&storagev1.StorageClass{},
+		&v1alpha1.AWSNodeTemplate{},
+	}
+)
+
+func (env *Environment) ExpectCreated(objects ...client.Object) {
+	for _, object := range objects {
+		object.SetLabels(map[string]string{EnvironmentLabelName: env.Options.EnvironmentName})
```
Comment (tzneal) [inline test/pkg/environment/expectations.go:34]: Shouldn't this merge the environment label instead of replacing existing?

File: test/pkg/environment/expectations.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+func (env *Environment) ExpectCleaned() {
+	namespaces := &v1.NamespaceList{}
+	env.Expect(env.Client.List(env, namespaces)).To(Succeed())
+	wg := sync.WaitGroup{}
+	for _, object := range CleanableObjects {
+		for _, namespace := range namespaces.Items {
+			wg.Add(1)
+			go func(object client.Object, namespace string) {
+				env.Expect(env.Client.DeleteAllOf(env, object,
```
Comment (tzneal) [inline test/pkg/environment/expectations.go:47]: Maybe add `&client.DeleteAllOfOptions{DeleteOptions: client.DeleteOptions{GracePeriodSeconds: ptr.Int64(0)}})`

File: test/pkg/environment/expectations.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+func (env *Environment) ExpectEventuallyHealthy() {
+	env.Eventually(func(g Gomega) error {
+		env.ExpectNodesHealthy(g)
+		env.ExpectPodsHealthy(g)
+		return nil
+	}).Should(Succeed())
+}
+
+func (env *Environment) ExpectNodesHealthy(g Gomega) {
```
Comment (tzneal) [inline test/pkg/environment/expectations.go:66]: Make this non-exported and same with `expectPodsHealthy`?

File: pkg/test/provisioner.go [hunk 4]
```diff
@@ -77,6 +73,14 @@ func Provisioner(overrides ...ProvisionerOptions) *v1alpha5.Provisioner {
 		},
 		Status: options.Status,
 	}
+
+	if options.ProviderRef == nil {
+		if options.Provider == nil {
+			options.Provider = struct{}{}
+		}
+		provider, _ := json.Marshal(options.Provider)
```
Comment (tzneal) [inline pkg/test/provisioner.go:81]: Shouldn't fail, but wouldn't hurt to `Expect(err).To(BeNil())`

File: test/pkg/environment/environment.go [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+	client, err := NewLocalClient()
+	if err != nil {
+		return nil, err
+	}
+	options, err := NewOptions()
+	if err != nil {
+		return nil, err
+	}
+	g := gomega.NewGomega(ginkgo.Fail)
+	g.SetDefaultEventuallyTimeout(10 * time.Minute)
```
Comment (tzneal) [inline test/pkg/environment/environment.go:44]: Nice! Didn't know this existed. Maybe do `g.SetDefaultEventuallyPollingInterval(1*time.Second)` as well, looks like the default is 10ms

File: test/pkg/environment/expectations.go (hunk 2 above)
Comment (ellistarn) [reply test/pkg/environment/expectations.go:47]: I wanted this to be closer to natural usage of clusters and include normal deletion flows rather than force deletion.

File: pkg/test/provisioner.go (hunk 4 above)
Comment (ellistarn) [reply pkg/test/provisioner.go:81]: Trying to keep this decoupled from gomega. I'll panic instead.

Comment (ellistarn) [reply pkg/test/provisioner.go:81]: We're also totally ignoring the validation 😭 
`utilruntime.Must(provisioner.Validate(context.Background()))`

File: test/tasks/before-karpenter-suite.yaml [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+      EOF
+  - name: kubeconfig
+    image: docker.io/njtran/scratch:latest
+    script: |
+        aws eks update-kubeconfig --name $(params.cluster-name)
+
+        TEMPOUT=$(mktemp)
+        curl -fsSL https://karpenter.sh/"$(params.karpenter-version)"/getting-started/getting-started-with-eksctl/cloudformation.yaml  > $TEMPOUT \
+        && aws cloudformation deploy \
+          --stack-name "Karpenter-$(params.cluster-name)" \
```
Comment (njtran) [inline test/tasks/before-karpenter-suite.yaml:46]: This is going to collide with our karpenter installation if using the same cluster name as the management cluster. Can we make it the following? It may be better to change up the host-cluster setup too so that we don't run into the chance of IAM Role name conflicts.
```suggestion
          --stack-name "KarpenterTesting-$(params.cluster-name)" \
```

File: test/tasks/go-suite.yaml [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+apiVersion: tekton.dev/v1beta1
+kind: ClusterTask
+metadata:
+  name: ginkgo-suite
+spec:
+  description: |
+    Execute a ginkgo suite on the provided package
+  params:
+    - name: repository
+    - name: path
```
Comment (njtran) [inline test/tasks/go-suite.yaml:11]: Can you put descriptions for these

**Review by njtran — COMMENTED** [review]
Let’s discuss how we want to separate tasks, clusterTasks, and pipelines in the filesystem. If someone’s importing all the files from GitHub this is quite easy and nice to find the file definitions in the repo as you've done it. 

As the amount of configs increase: if someone’s doing this manually, they have to figure out how to find all related tasks then apply them manually. I think we should keep ClusterTasks in their own ClusterTasks folder and then any tasks in their respective suites. WDYT?

File: test/infrastructure/scripts/step-02-eksctl-cluster.sh [hunk 8]
```diff
@@ -14,7 +14,11 @@ managedNodeGroups:
     name: ${CLUSTER_NAME}-system-pool
     desiredCapacity: 1
     minSize: 1
-    maxSize: 10
+    maxSize: 1
+    taints:
```
[the remark this answers is not in the document]
Comment (ellistarn) [reply test/infrastructure/scripts/step-02-eksctl-cluster.sh:18]: Can we do this separately? I didn't change he size of the nodegroup, just changed the max, which was strangely set to a high value. The only reason to have a max this high is if you're doing some sort of autoscaling on this group (which we're not). If we decide to manually scale this group in the UI, you can easily change max and desired. If we decide to change the size of this group in IAC, then the change should live in IAC.

Put simply, every value you configure has intent, and more apparent you can make that intent, the better. Having maxsize != desiredsize implies that we're going to autoscale this group -- which we aren't.

[the remark this answers is not in the document]
Comment (njtran) [reply test/infrastructure/scripts/step-02-eksctl-cluster.sh:18]: Sure. I think Brandon addressed this in his PR. Happy to defer to his, and I agree with the != maxsize point.

File: test/infrastructure/scripts/step-02-eksctl-cluster.sh [hunk 9]
```diff
@@ -1,4 +1,4 @@
-eksctl create cluster -f - << EOF
+eksctl create cluster -f - <<EOF
```
[the remark this answers is not in the document]
Comment (ellistarn) [reply test/infrastructure/scripts/step-02-eksctl-cluster.sh:1]: I have a bash linter/formatter in vscode that's cleaning this up. Do you have one with different opinions?

File: test/infrastructure/scripts/step-06-karpenter.sh [hunk 10]
```diff
@@ -21,15 +21,16 @@ export KARPENTER_IAM_ROLE_ARN="arn:aws:iam::${AWS_ACCOUNT_ID}:role/${CLUSTER_NAM
...
 
-if [ -z "$(helm repo list | grep karpenter)" ] ; then
+if [ -z "$(helm repo list | grep karpenter)" ]; then
   helm repo add karpenter https://charts.karpenter.sh
 fi
 helm repo update
-helm upgrade --install --namespace karpenter --create-namespace \
-  karpenter karpenter/karpenter \
-  --version ${KARPENTER_VERSION} \
+helm upgrade --install --namespace karpenter --create-namespace karpenter --version v0.13.1 \
```
[the remark this answers is not in the document]
Comment (ellistarn) [reply test/infrastructure/scripts/step-06-karpenter.sh:28]: This is a really bad precedent to set. We can't be behind on karpenter in karpenter testing. Can't we set a higher ttl?

[the remark this answers is not in the document]
Comment (ellistarn) [reply test/infrastructure/scripts/step-06-karpenter.sh:28]: > Vending the Karpenter installation for KIT can be possible in the future and where they would establish the Karpenter version we could use.

Is this something like kitctl?

File: test/infrastructure/scripts/step-07-provisioners.sh [hunk 11]
```diff
@@ -15,17 +15,17 @@ spec:
     values: ["amd64"]
   ttlSecondsAfterEmpty: 180
   provider:
-    instanceProfile: KarpenterGuestClusterNodeInstanceProfile-${CLUSTER_NAME}
+    instanceProfile: KarpenterNodeInstanceProfile-${CLUSTER_NAME}
```
[the remark this answers is not in the document]
Comment (ellistarn) [reply test/infrastructure/scripts/step-07-provisioners.sh:18]: Good point.

File: test/infrastructure/scripts/step-02-eksctl-cluster.sh (hunk 9 above)
[the remark this answers is not in the document]
Comment (njtran) [reply test/infrastructure/scripts/step-02-eksctl-cluster.sh:1]: Nope. Looks like I gotta get one for myself :)

File: test/infrastructure/scripts/step-07-provisioners.sh [hunk 12]
```diff
@@ -40,5 +40,5 @@ spec:
     subnetSelector:
       karpenter.sh/discovery: ${CLUSTER_NAME}
     securityGroupSelector:
-      kubernetes.io/cluster/${CLUSTER_NAME}: owned
+      karpenter.sh/discovery/${CLUSTER_NAME}: owned
```
[the remark this answers is not in the document]
Comment (ellistarn) [reply test/infrastructure/scripts/step-07-provisioners.sh:43]: How did I miss this.

File: test/infrastructure/scripts/step-06-karpenter.sh (hunk 10 above)
[the remark this answers is not in the document]
Comment (njtran) [reply test/infrastructure/scripts/step-06-karpenter.sh:28]: Possibly. If this is an issue with pre-binding then I want to be able to figure this out. This was a fairly large change in assumptions for Karpenter so I don't think we'll be sticking here for long. We use a TTL of 180 right now.

File: test/infrastructure/scripts/step-02-eksctl-cluster.sh (hunk 8 above)
[the remark this answers is not in the document]
Comment (bwagner5) [reply test/infrastructure/scripts/step-02-eksctl-cluster.sh:18]: Generally, you want maxSize to be 2x desired for upgrades. I'm not sure if MNG raises that for you when it does an upgrade?

File: test/suites/integration/pipeline.yaml [hunk 13]
```diff
@@ -0,0 +1,20 @@
+---
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: karpenter-integration-suite
+spec:
+  params:
+    - name: commit
```
[the remark this answers is not in the document]
Comment (ellistarn) [reply test/suites/integration/pipeline.yaml:8]: I think it's self documenting as is.

File: test/suites/integration/pipeline.yaml [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+  tasks:
+  - name: github-action
+    taskRef:
+      name: github-action
```
[the remark this answers is not in the document]
Comment (ellistarn) [reply test/suites/integration/pipeline.yaml:12]: Not a big fan of self naming objects. `deployments/foo-deployment` stutters

File: pkg/test/pods.go [hunk 15]
```diff
@@ -93,7 +93,7 @@ func Pod(overrides ...PodOptions) *v1.Pod {
 			Tolerations:               options.Tolerations,
 			InitContainers: []v1.Container{{
 				Name:      strings.ToLower(sequentialRandomName()),
-				Image:     options.Image,
+				Image:     "alpine",
```
[the remark this answers is not in the document]
Comment (ellistarn) [reply pkg/test/pods.go:96]: Pause doesn't work as an init container. Needed something that could exit.

File: pkg/test/provisioner.go [hunk 16]
```diff
@@ -59,25 +61,35 @@ func Provisioner(overrides ...ProvisionerOptions) *v1alpha5.Provisioner {
...
 		Spec: v1alpha5.ProvisionerSpec{
 			Requirements:         options.Requirements,
 			KubeletConfiguration: options.Kubelet,
-			Provider:             &runtime.RawExtension{Raw: provider},
 			ProviderRef:          options.ProviderRef,
 			Taints:               options.Taints,
 			StartupTaints:        options.StartupTaints,
 			Labels:               options.Labels,
 			Limits:               &v1alpha5.Limits{Resources: options.Limits},
+			TTLSecondsAfterEmpty: ptr.Int64(10),
```
[the remark this answers is not in the document]
Comment (ellistarn) [reply pkg/test/provisioner.go:75]: Eventually sure -- I wanted to default for 10 until we need to override.

File: pkg/test/provisioner.go [hunk 17]
```diff
@@ -59,25 +61,35 @@ func Provisioner(overrides ...ProvisionerOptions) *v1alpha5.Provisioner {
...
+		provider, err := json.Marshal(options.Provider)
+		if err != nil {
+			panic(err)
+		}
+		provisioner.Spec.Provider = &runtime.RawExtension{Raw: provider}
+	}
 	provisioner.SetDefaults(context.Background())
-	_ = provisioner.Validate(context.Background())
+	if err := provisioner.Validate(context.Background()); err != nil {
+		logging.FromContext(context.TODO()).Info("TODO: Fix the tests that cause this")
```
[the remark this answers is not in the document]
Comment (ellistarn) [reply pkg/test/provisioner.go:92]: A ton of existing tests break if I fail here, since the error was originally squashed incorrectly. I wanted to log and clean it up later, since I'm OOTO for a few days and wanted to unblock folks.

File: test/infrastructure/scripts/step-01-config.sh [hunk 18]
```diff
@@ -1,6 +1,5 @@
-export CLUSTER_NAME="${CLUSTER_NAME:-karpenter-test-cluster}"
+export CLUSTER_NAME="${CLUSTER_NAME:-karpenter-test-infrastructure}"
```
[the remark this answers is not in the document]
Comment (ellistarn) [reply test/infrastructure/scripts/step-01-config.sh:1]: `karpenter-test-management` is less idiomatic than `karpenter-test-infrastructure` to me.

**Review by njtran — CHANGES_REQUESTED** [review]
I think we have some disagreements on what the CI Infra right now. I think the ginkgo work looks great and am happy to approve that if we scope this one down.

File: test/pkg/environment/expectations.go [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+	storagev1 "k8s.io/api/storage/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/aws/karpenter/pkg/apis/awsnodetemplate/v1alpha1"
+	"github.com/aws/karpenter/pkg/apis/provisioning/v1alpha5"
+)
+
+var (
+	EnvironmentLabelName = "testing.karpenter.sh/environment-name"
+	CleanableObjects     = []client.Object{
```
Comment (njtran) [inline test/pkg/environment/expectations.go:20]: How do we ensure that Nodes are cleaned up? I think the TTL on the Provisioners is meant to do that, but if a test fails to delete nodes (which could happen), then we want our Cleanup to also include those.

File: test/suites/integration/suite_test.go [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+	RunSpecs(t, "Integration Suite")
+}
+
+var _ = AfterEach(func() {
+	env.ExpectCleaned()
+})
+
+var _ = Describe("Sanity Checks", func() {
+	It("should provision nodes", func() {
+		provider := test.AWSNodeTemplate(test.AWSNodeTemplateOptions{AWS: v1alpha1.AWS{
```
Comment (njtran) [inline test/suites/integration/suite_test.go:32]: If the Karpenter being tested here doesn't have a defaultInstanceProfile, this should fail. I think it's fine to leave as is, but I think we should make sure we can test both modes eventually.

**Review by njtran — COMMENTED** [review]
Nice! Just a nit or two

File: test/pkg/environment/expectations.go (hunk 19 above)
Comment (njtran) [reply test/pkg/environment/expectations.go:20]: Ah I forgot that Nodes are owned by provisioners now, so this should be good.

Comment (ellistarn) [reply test/pkg/environment/expectations.go:20]: We definitely need to test scale down, too. I think this library will need to be greatly expanded upon (e.g. wait for things to cleanup). This is just to get us started.

**Review by njtran — APPROVED** [review]
